File: src/tinymonitor/alerts/ntfy.py
import requests
from datetime import datetime
from tinymonitor.alerts.provider import AlertProvider, logger
from tinymonitor.utils import system_info

class NtfyAlert(AlertProvider):
    def send(self, component, level, value, title, message):
        topic_url = self.config.get("topic_url")
        if not topic_url:
            logger.error(f"[{self.name}] No topic_url provided")
            return

        # Ntfy priority mapping (1=min, 3=default, 5=max)
        if level == "CRITICAL":
            priority = 5
            tags = "rotating_light,critical"
        elif level == "WARNING":
            priority = 3
            tags = "warning"
        else:
            priority = 1
            tags = "information_source"

        # System Info
        hostname = system_info.get_hostname()
        execution_time = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        ip_private = system_info.get_private_ip()
        ip_public = system_info.get_public_ip()
        load_avg = system_info.get_load_avg()
        uptime_pretty = system_info.get_uptime()
        
        # Enriched message construction (Markdown supported by ntfy)
        full_message = (
            f"**Component** : {component}\n"
            f"**Value**     : {value}\n"
            f"**Level**     : {level}\n\n"
            f"__Machine Context__\n"
            f"🖥️ **Server**    : `{hostname}`\n"
            f"🏠 **Private IP**: `{ip_private}`\n"
            f"🌍 **Public IP** : `{ip_public}`\n"
            f"⚙️ **Load Avg**  : `{load_avg}`\n"
            f"⏱️ **Uptime**    : `{uptime_pretty}`\n"
            f"🕒 **Time**      : {execution_time}"
        )

        headers = {
            "Title": title,
            "Priority": str(priority),
            "Tags": tags,
            "Markdown": "yes"
        }

        # Authentication handling if token is provided
        token = self.config.get("token")
        if token:
            headers["Authorization"] = f"Bearer {token}"

        try:
            response = requests.post(
                topic_url,
                data=full_message.encode('utf-8'),
                headers=headers,
                timeout=10
            )
            if response.status_code == 200:
                logger.info(f"[{self.name}] Alert sent successfully")
            else:
                logger.error(f"[{self.name}] Failed to send alert: {response.text}")
        except Exception as e:
            logger.error(f"[{self.name}] Error sending alert: {e}")

File: internal/alerts/ntfy.go
package alerts

import (
	"bytes"
	"fmt"
	"net/http"
	"time"

	"github.com/Gu1llaum-3/tinymonitor/internal/config"
	"github.com/Gu1llaum-3/tinymonitor/internal/models"
	"github.com/Gu1llaum-3/tinymonitor/internal/utils"
)

// NtfyProvider sends alerts to Ntfy
type NtfyProvider struct {
	BaseProvider
	topicURL string
	token    string
}

// NewNtfyProvider creates a new Ntfy provider
func NewNtfyProvider(cfg config.NtfyConfig) *NtfyProvider {
	return &NtfyProvider{
		BaseProvider: BaseProvider{
			ProviderName: "ntfy",
			Enabled:      cfg.Enabled,
			Levels:       cfg.Levels,
			Rules:        cfg.Rules,
		},
		topicURL: cfg.TopicURL,
		token:    cfg.Token,
	}
}

// Send sends an alert to Ntfy
func (p *NtfyProvider) Send(alert models.Alert) error {
	if p.topicURL == "" {
		return fmt.Errorf("no topic_url provided")
	}

	// Priority mapping
	var priority string
	var tags string
	switch alert.Level {
	case models.SeverityCritical:
		priority = "5"
		tags = "rotating_light,critical"
	case models.SeverityWarning:
		priority = "3"
		tags = "warning"
	case models.SeverityRecovery:
		priority = "2"
		tags = "white_check_mark,recovered"
	default:
		priority = "1"
		tags = "information_source"
	}

	// System Info
	hostname := utils.GetHostname()
	executionTime := time.Now().Format("2006-01-02 15:04:05")
	ipPrivate := utils.GetPrivateIP()
	ipPublic := utils.GetPublicIP()
	loadAvg := utils.GetLoadAvg()
	uptimePretty := utils.GetUptime()

	// Enriched message (Markdown supported)
	fullMessage := fmt.Sprintf(`**Component** : %s
**Value**     : %s
**Level**     : %s

__Machine Context__
🖥️ **Server**    : `+"`%s`"+`
🏠 **Private IP**: `+"`%s`"+`
🌍 **Public IP** : `+"`%s`"+`
⚙️ **Load Avg**  : `+"`%s`"+`
⏱️ **Uptime**    : `+"`%s`"+`
🕒 **Time**      : %s`,
		alert.Component, alert.Value, alert.Level,
		hostname, ipPrivate, ipPublic, loadAvg, uptimePretty, executionTime)

	req, err := http.NewRequest("POST", p.topicURL, bytes.NewReader([]byte(fullMessage)))
	if err != nil {
		return err
	}

	req.Header.Set("Title", alert.Title)
	req.Header.Set("Priority", priority)
	req.Header.Set("Tags", tags)
	req.Header.Set("Markdown", "yes")

	if p.token != "" {
		req.Header.Set("Authorization", "Bearer "+p.token)
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("failed to send alert: status %d", resp.StatusCode)
	}

	LogInfo(p.ProviderName, "Alert sent successfully")
	return nil
}
